# GES-5 flank analysis workflow

## Data curation

- We retrived 1,375 contigs from NCBI's Pathogen Detection [MicroBIGG-E](https://www.ncbi.nlm.nih.gov/pathogens/microbigge) using the query `element_symbol:blaGES*` (as of 27/02/2023). The full metadata table was also downloaded.
- Contigs were annotated for GES-5 using the NCBI GES-5 protein reference sequence [WP_012658785.1](https://www.ncbi.nlm.nih.gov/protein/WP_012658785.1) and blastx. We filtered by 100% identity/coverage and a single hit, giving 431 contigs.
- Contigs were then deduplicated as follows: first, we took the longest contig from each BioSample, discarding the rest. In the case of ties, we chose a random representative. Then, within each BioProject, we discarded any contigs that were perfectly contained in another. If this resulted in all contigs from a BioProject being discarded, we chose a random, longest representative.
- The script `deduplication.R` was used to wrangle the various inputs (NCBI metadata, contig lengths, BioProject accessions, mash screen) then deduplicate the contigs.

### Contig lengths

Contig sequence lengths were calculated using `fastaLengths.py`. Usage is `python3 fastaLengths.py contigs.fa` which outputs `lengths.tsv`. Requires `SimpleFastaParser` from [Biopython](https://github.com/biopython/biopython).

### BioProject accessions

BioProject accessions were retreived using NCBI's [E-utilities](https://www.ncbi.nlm.nih.gov/books/NBK179288/). The following code writes a tab-delimited table of BioProjects for a given list of BioSamples:
```
cat BioSamples.txt | { 
while read line
do
echo "$line" | tr '\n' '\t'
esearch -db biosample -query $line < /dev/null | elink -target bioproject | efetch -format docsum | xtract -pattern DocumentSummary -element Project_Acc | tr '\n' '\t' && echo ""
sleep 1
done
} > BioProjects.tsv
```
BioProject accessions were manually curated to remove any meta-analyses which reused contigs from earlier studies. 

### Sequence containment
Pairwise contig containment used [Mash](https://github.com/marbl/Mash) (v. 2.2):

```
mash sketch -s 1000000 ./contigs/*.fasta -o contigs
echo contigs/*.fasta | xargs -n 1 mash screen contigs.msh > mash-output.tsv
```
With this command, the Mash output will only list the 'screened' contigs, not the contig that we are scoring containment in. For example,
```
x
a.fasta
b.fasta
c.fasta
a.fasta
b.fasta
c.fasta
a.fasta
b.fasta
c.fasta
```
really means scoring the containment of x in y:
``` 
x y
a.fasta a.fasta
b.fasta a.fasta
c.fasta a.fasta
a.fasta b.fasta
b.fasta b.fasta
c.fasta b.fasta
a.fasta c.fasta
b.fasta c.fasta
c.fasta c.fasta
```
A sense check here is that given *m* contigs, every *(n-1)m+n*-th row should score 1 since we are screening a contig against itself.

## Extracting flanking sequences

- Flanking sequences were extracted using the python tool [Flanker](https://github.com/wtmatlock/flanker) with a custom [Abricate](https://github.com/tseemann/abricate) database containing only the NCBI GES-5 nucleotide reference sequence [NG_049137.1](https://www.ncbi.nlm.nih.gov/nuccore/NG_049137.1). Flanker was run by the Slurm scheduler script `runFlanker.sh` within the Conda environment described [here](https://flanker.readthedocs.io/en/latest/#installation). For a contig called `foo.fasta` and `--window bar`, the output file will be called `foo_GES-5_bar_both_flank.fasta`.

## NCBI annotations

Contig annotations were pre-generated by NCBI's [Prokaryotic Genome Annotation Pipeline](https://www.ncbi.nlm.nih.gov/genome/annotation_prok/), and retrived as follows:

```
mkdir gffs
cat contigAccessions.txt | { 
while read line
do
wget "https://www.ncbi.nlm.nih.gov/sviewer/viewer.cgi?db=nuccore&report=gff3&id=${line}" -O ./gffs/"$line".gff
sleep 1
done
}
```
## Integron-specific annotations

Additional, integron-specific annoations were generated using [Integron Finder](https://github.com/gem-pasteur/Integron_Finder) (v. 2.0.2) with the command
```
integron_finder ./contigs/*.fasta --local-max --promoter-attI --lin
```
For each contig, this wrote a `.integron` file, which we combined using
```
awk 'FNR==1 && NR!=1{next;}{print}' ./*.integrons > integron-finder-output.tsv
```

## Transposasable element annotations

We superseded NCBI's transposable element annotations with those generated from the ISfinder database and Abricate. Commands are given in the script `runISfinder.sh`.

## Integrase SNV profiles

The start/end positions and strand of the integrase annotations were extracted from the `.integron` files using
```
awk -F'\t' -v OFS='\t' 'NR>1 && $9=="intI" {print $2, $4, $5, $^}' ./*.integrons > integrase-positions.tsv
```
Since the full contigs were used, and not the GES-5 flanking sequences, some contained multiple *intI* annotations. These were manually removed at this stage by comparing integrase and GES-5 annotation positios. We extracted the integrase sequences using `extractIntegrase.py`. The Usage is `python3 extractIntegrase.py integrase-positions.tsv` which outputs individual files `foo-integrase.fasta` and the translated `foo-integrase-prot.fasta`. NB, the directory paths might need updating in the script.

Integrase sequences were aligned using [ClustalW](http://www.clustal.org/clustal2/) (v. 2.1) on slow/accurate mode. The SNP distance matrix was generated using [snp-dists](https://github.com/tseemann/snp-dists) (v. 0.8.2) with default parameters, and SNP constant sites were found using [snp-sites](https://github.com/sanger-pathogens/snp-sites) (v. 2.5.1) with `-C` and otherwise default paramaters. The SNP clustering/heatmap was generated using the script `plotHeatmap.R`.
